### ISSUE TITLE
Raised mod incompatibility with MekaSuit HUD

### ISSUE THREAD

**Issue by JohnGoten** [mekanism/Mekanism#8268]
### Issue description

Originally discovered in a personal instance (modpack development) of mine, I noticed after having configured the Raised mod to have a more lifted, bedrock-esque, hotbar visual, equipping a MekaSuit would have its HUD moved up by the equal amount as it was cut off at the top already, normally sitting comfortably within what's visible on screen
I was able to recreate this behavior within an isolated instance only using Mekanism and Raised, not even Sodium (+Iris) present. I also created an issue https://github.com/yurisuika/Raised/issues/100 and was shortly informed on their discord of them seemingly only hooking into the existing GUI layers present. Of course given the clearly unintended behavior, I can only assume the MekaSuit HUD is tied to that of the hotbar which should instead be disconnected and resolve this issue (fingers crossed).
![2024-12-07_15 13 32](https://github.com/user-attachments/assets/e54f2a7f-d336-4607-948a-5d2c31ee8311)

### Steps to reproduce

1. Equip a full set of MekaSuit armor
2. Access the Raised config via "`"
3. Move the hotbar (top left has options for chat etc. which are unaffected) on the Y axis using the top right slider
4. Watch as the HUD moves off-screen

### Minecraft version

1.21.1 (Latest)

### NeoForge version

21.1.84

### Mekanism version

10.7.7 (Latest)

### Other relevant versions

Raised 4.0.1 (Latest 1.21.1 as of now) https://modrinth.com/mod/raised/version/NeoForge-1.21.1-4.0.1

### If a (crash)log is relevant for this issue, link it here: (It's almost always relevant)

https://mclo.gs/IjUkWxk

**Comment by JohnGoten** [comment]
Thanks to yurisuika's help, never previously dealt with IntelliJ as your average consumer, I did manage to change the ClientRegistration.java line 318
![image](https://github.com/user-attachments/assets/2890fdbc-6e36-4ecb-b05d-8efdb71ca54d)


`        event.registerAbove(VanillaGuiLayers.HOTBAR, Mekanism.rl("hud"), MekanismHUD.INSTANCE);`

into

`       event.registerAbove(VanillaGuiLayers.CROSSHAIR, Mekanism.rl("hud"), MekanismHUD.INSTANCE);`

as this seemed to be the most reasonable option available, although I tried

`        event.registerAboveAll(Mekanism.rl("hud"), MekanismHUD.INSTANCE);`

covering up chat and the F3 menu, probably bad idea, same for BelowAll as yurisuika hinted at me regarding just the vanilla frozen or portal overlay whilst attaching it to the crosshair keeps it behind all those whilst unaffected by what appears to be any and all of Raised's configuration options, effectively solving the problem I was dealing with

and please tell me if there's a better way to demonstrate this, literally first time going this "deep" into the matter


EDIT: setting it below the SLEEP_OVERLAY may be the smartest choice available and a direct recommendation from yurisuika
![image](https://github.com/user-attachments/assets/72c7be5b-4de6-4c5b-83cd-1f294e844f2b)

**Comment by thiakil** [comment]
I believe this is on Raised to fix. Layers are named with `ResourceLocation`s and therefor they should not be moving layers that aren't the intended target.

Crosshair does not seem like the correct layer to target it to either, but glad it works for your use case.

**Comment by JohnGoten** [comment]
That is why I mentioned the sleep overlay part, past that point it's their business what to make of this information

**Comment by yurisuika** [comment]
What I was trying to explain is that the Forge GUI registry doesn't determine when a modded GUI element renders, it is only determining where it renders in the layers. Going by the vanilla layers, hotbar is near the bottom with crosshair one below it. So if you're registering your GUI below the hotbar, it will be above the crosshair. In realistic usage in this mod's case, no matter how much you squish the screen around, you still can't get any of the GUI stuff to intersect with the crosshair. So, if you were to register it below the crosshair, it would make no difference visually. Any vanilla stuff is going to render above it, like the boss bars presently do if your window is narrow enough.

In order to support mod GUIs that use the register event on Forge instead of the render event, Raised captures any modded HUD element's resource location that is "associated" with the vanilla layers it must translate. The only way to make this association that I am aware of is to determine for which vanilla layer it is registered above/below. Effectively, registering above the crosshair and registering below the hotbar does the exact same thing, that is until you have other mods doing one or the other and then you gotta get the event priority involved. Hence my suggestion, as it would resolve the conflict and should not realistically impact anything visually.

As far as the sleep overlay goes, I was suggesting that a GUI element should be placed no higher than that, unless you want it clearly visible when you sleep. Like the crosshair, Raised does not interact with the sleep overlay layer in any form.

The only reasonable non-hardcoded way to make a change on my end would be to add some sort of Forge exclusive blacklist that a user could configure with a command. This would still rely on the user knowing that the blacklist exists and understanding how to use it. Users may still end up filing reports because of this. Would be leagues better than having users have to manually add the modded registered overlays to the Raised overlay groups in a config, but still not without a substantial loss in ease-of-use.

However, changing the mekanism hud to be associated with a different element (such as crosshair) would eliminate any conflict and have no negative results. Register above crosshair, below crosshair, above camera overlays, it should all visually produce the same result w/o conflict w/ Raised.

For reference, these are the elements that Raised may translate. Of course, the toasts do not have a Forge encapsulation. https://github.com/yurisuika/Raised/blob/NeoForge-1.21.1/src/main/java/dev/yurisuika/raised/util/Overlay.java

**Comment by thiakil** [comment]
> the Forge GUI registry doesn't determine when a modded GUI element renders, it is only determining where it renders in the layers

That's not correct, it determines the order they render in. There are even events for before and after a layer has rendered. See `net.neoforged.neoforge.client.gui.GuiLayerManager`

There should be no need to capture anything related to other gui layers. The pre/post render events have the layer name supplied, so you could do your translations there.

**Comment by JohnGoten** [comment]
At the end of the day changing the variable out retains the same implementation but set to a different GUI layer as mentioned with my original post
This argument could go on forever based on opinions or whatever but the end result is likely to be changing parts of this
`        event.registerAbove(VanillaGuiLayers.HOTBAR, Mekanism.rl("hud"), MekanismHUD.INSTANCE);` to `registerBelow(VanillaGuiLayers.SLEEP_OVERLAY` as suggested speaking to them on their discord server
it could also merely be swapping `VanillaGuiLayers.HOTBAR` to `VanillaGuiLayers.CROSSHAIR` or whichever one really ends up fitting the best whilst not conflicting with the other, simple as that

To me this sounds most logical and does not require much of an argument to get done as even after personal testing running the mod set to `VanillaGuiLayers.CROSSHAIR` for the sake of proving a point with nothing to report other than it now not being affected by the Raised mod ¯\\\_(ツ)\_/¯


Not to mention it wasn't really any different with another different case of this sorts using https://modrinth.com/mod/camerapture which was resolved by very much doing the same thing but in Mekanism's case even easier given the existing code only needing a word or two swapped
https://github.com/chrrs/camerapture/issues/35
https://github.com/chrrs/camerapture/commit/f76beb3b88ae987072f563baa168f6e9b95e5746

**Comment by thiakil** [comment]
@yurisuika <https://github.com/yurisuika/Raised/blob/1dd2090c1a0ee434b9f47b738efc78a68823401c/src/main/java/dev/yurisuika/raised/mixin/client/event/RegisterGuiLayersEventMixin.java#L28>

You seem to be making the assumption that if a particular is used _at all_ in positioning, then it's part of that section, which seems very very wrong. Above maaaaybe (preferably not), but definitely not _below_

I believe this should be opt-in for modded GUI additions, or provide users a way to move the named layers individually. 
While Mekanism could make a change on our side, this is going to keep happening for you if you don't change those mixins.

**Comment by pupnewfster** [comment]
> What I was trying to explain is that the Forge GUI registry doesn't determine when a modded GUI element renders, it is only determining where it renders in the layers.

It is possible to get what layer is being rendered based on the [name exposed in the event](https://github.com/neoforged/NeoForge/blob/1.21.1/src/main/java/net/neoforged/neoforge/client/event/RenderGuiLayerEvent.java#L50). Strictly speaking yes to determine what layer it is after you would need to keep track of the order they are called, but a more proper solution would be to request that Neo exposes the list of registered layers (`GuiLayerManager.layers` or at least the `ResourceLocation` part of that list).

> Any vanilla stuff is going to render above it, like the boss bars presently do if your window is narrow enough.

This is intended.


> I believe this is on Raised to fix. Layers are named with `ResourceLocation`s and therefor they should not be moving layers that aren't the intended target.

I agree that the proper solution should be to not just automatically batch things into what general layer was rendered before or after it, as the pose stack that is used to render each layer (logically) does not leak any data in vanilla, as each layer is meant to be rendered individually. If you do want to make it so that it groups it by default, I would recommend exposing a list to players when configuring the position of "all" things that are considered a hotbar by raised, so that they can just opt-out/in of having the specific layer be modified with the hotbar's settings rather than having to apply the settings manually.

**Comment by yurisuika** [comment]
Yep, it wasn't that I thought that is what it meant, but rather that I had to make an assumption because of the limitations of how Raised was set up with an enum for the layers which were configurable. Coming soon it will instead use a map of all registered layers and will eliminate this incompatibility.
